pallets/omnipool/src/lib.rs

```diff
@@ -0,0 +1,1459 @@
+//! * `add_liquidity` - Adds liquidity of selected asset to the pool. Mints corresponding position NFT.
+//! * `remove_liquidity` - Removes liquidity of selected position from the pool. Partial withdrawals are allowed.
+//! * `sell` - Trades an asset in for asset out by selling given amount of asset in.
+//! * `buy` - Trades an asset in for asset out by buying given amount of asset out.
+//! * `set_asset_tradable_state` - Updates asset's tradable asset with new flags. This allows/forbids asset operation such SELL,BUY,ADD or  REMOVE liquidtityy.
+//! * `refund_refused_asset` - Refunds the initial liquidity amount sent to pool account prior to add_token if the token has been refused to be added.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet_prelude::{DispatchResult, Get};
+use frame_support::require_transactional;
+use frame_support::PalletId;
+use frame_support::{ensure, transactional};
+use sp_runtime::traits::{AccountIdConversion, AtLeast32BitUnsigned, One};
+use sp_runtime::traits::{CheckedAdd, CheckedSub, Zero};
+use sp_std::ops::{Add, Sub};
+use sp_std::prelude::*;
+
+use frame_support::traits::tokens::nonfungibles::{Create, Inspect, Mutate};
+use hydra_dx_math::omnipool::types::BalanceUpdate;
+use hydradx_traits::Registry;
+use orml_traits::MultiCurrency;
+use sp_runtime::{ArithmeticError, DispatchError, FixedPointNumber, FixedU128, Permill};
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+mod benchmarks;
+
+#[cfg(test)]
+mod tests;
+
+mod types;
+pub mod weights;
+
+use crate::types::{AssetReserveState, AssetState, Balance, Price, SimpleImbalance, Tradability};
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+/// NFT class id type of provided nft implementation
+type NFTClassIdOf<T> = <<T as Config>::NFTHandler as Inspect<<T as frame_system::Config>::AccountId>>::ClassId;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use crate::types::{Position, Price, SimpleImbalance, Tradability};
+	use codec::HasCompact;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use hydra_dx_math::omnipool::types::BalanceUpdate;
+	use sp_runtime::ArithmeticError;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(crate) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Identifier for the class of asset.
+		type AssetId: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ HasCompact
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen
+			+ TypeInfo;
+
+		/// Multi currency mechanism
+		type Currency: MultiCurrency<Self::AccountId, CurrencyId = Self::AssetId, Balance = Balance>;
+
+		/// Add token origin
+		type AddTokenOrigin: EnsureOrigin<Self::Origin>;
+
+		/// Origin to be able to change asset tradabilit state and initialize Omnipool.
+		type ManageOrigin: EnsureOrigin<Self::Origin>;
```

Comment:
`TechnicalOrigin` would be maybe better name for this, manage is way too broad, and implies control, we are not managing anyone assets